### PR TITLE
Allow modules from outside the Arkouda source dir to be built into the server

### PR DESCRIPTION
**Pull request by bmcdonald3**
Today, the modular Arkouda build only works for modules in the Arkouda `src/` directory, but this PR enables users to specify modules from any directory by supplying the absolute path in an environment variable called `ARKOUDA_SERVER_USER_MODULES`.

This means a user could add `TestMsg` module name to the `ServerModules.cfg` file and then build Arkouda with the command `ARKOUDA_SERVER_USER_MODULES=/path/to/TestMsg.chpl make` and that module would be included in the build. The path in the environment variable is appended onto the `chpl` command so that the compiler can find the file.

Resolves: https://github.com/Bears-R-Us/arkouda/issues/1046

### CODE REVIEW

File: src/serverModuleGen.py [hunk 1]
```diff
@@ -22,6 +25,21 @@ def generateServerIncludes(config_filename, reg_filename):
 
     serverfile.write("}\n")
 
+def generateSearchPaths(filename):
+    with open(filename) as configfile:
+        modules = configfile.readlines()
+        ret = ""
+        for module in modules:
+            module = (module.rstrip()).lstrip()
```
Comment (glitch) [inline src/serverModuleGen.py:33]: I think you can simplify this as `module = module.strip()`  which should trim whitespace from both ends.

File: src/serverModuleGen.py [hunk 2]
```diff
@@ -22,6 +25,21 @@ def generateServerIncludes(config_filename, reg_filename):
...
+            # Represents a path to a file
+            if len(module) > 0 and module[0] == '/':
```
Comment (glitch) [inline src/serverModuleGen.py:35]: You can simplify this as `if module.startswith("/")`

Comment (bmcdonald3) [reply src/serverModuleGen.py]: Thanks for the suggestions, both are better for sure!

**Review by glitch — APPROVED** [review]

File: MODULAR.md [hunk 3]
```diff
@@ -12,9 +12,7 @@ New config files must follow the same format, listing one module name per line w
 
 ### Adding new modules into the build process
 
-Adding a custom module into the Arkouda buld process is treated the same way as a regular module: added in a single line in the server configuration file.
-
-Currently, modules will only be found by the Arkouda build if located in the `src/` directory.
+The only unique aspect of adding modules from other directories into the Arkouda build process is that a complete path must be specified for all modules not found in the Arkouda `src/` directory. This would actually be done by adding a line such as `/Users/test-user/path/to/mod` to `ServerModules.cfg` to add the module `mod.chpl` to the Arkouda build. Aside from that, adding a custom module into the Arkouda buld process is treated the same way as a regular module: added in a single line in the server configuration file.
```
Comment (pierce314159) [inline MODULAR.md:15]: very small note:
I think it's clearer here to say "This is done by" instead of "This would actually be done by"

**Review by pierce314159 — APPROVED** [review]
Looks good to me!

File: MODULAR.md (hunk 3 above)
Comment (bmcdonald3) [reply MODULAR.md]: Makes sense, thanks for pointing that out; added.

**Comment by mhmerrill** [comment]
IMO this is a step forward but I'm a little taken back by the need to spec absolute paths. Maybe we could also spec with an env var or something like `$OTHER_PKG/PkgMsg.chpl` or something? make sense? I am not sure I want my local directory structure exposed if I gave the module list to someone else...

**Comment by bmcdonald3** [comment]
Ya, that makes sense to me, I think having an environment variable to specify the direct paths is a better idea, I can work on switching over to that.

**Comment by glitch** [comment]
A minor request to have the env var prefixed with `ARKOUDA_`, so it limits any potential clash with a users environment (and makes it easy to `env|grep ARK` and dump out all of the arkouda variables.

**Comment by mhmerrill** [comment]
> A minor request to have the env var prefixed with `ARKOUDA_`, so it limits any potential clash with a users environment (and makes it easy to `env|grep ARK` and dump out all of the arkouda variables.

is that something we should enforce? or a usage convention?

**Comment by glitch** [comment]
I'm realizing we could really leave it in the user's hands.  In theory they can use whatever env var or directory they want right?

**Comment by mhmerrill** [comment]
> I'm realizing we could really leave it in the user's hands. In theory they can use whatever env var or directory they want right?

i think so... maybe they have a chapel package managed by mason and it might also have an Arkouda extension...

**Review by pierce314159 — APPROVED** [review]

File: MODULAR.md [hunk 4]
```diff
@@ -12,9 +12,7 @@ New config files must follow the same format, listing one module name per line w
 
 ### Adding new modules into the build process
 
-Adding a custom module into the Arkouda buld process is treated the same way as a regular module: added in a single line in the server configuration file.
-
-Currently, modules will only be found by the Arkouda build if located in the `src/` directory.
+The only unique aspect of adding modules from other directories into the Arkouda build process is that a complete path must be specified for all modules not found in the Arkouda `src/` directory. This is done by adding a line such as `/Users/test-user/path/to/mod` to `ServerModules.cfg` to add the module `mod.chpl` to the Arkouda build. Aside from that, adding a custom module into the Arkouda buld process is treated the same way as a regular module: added in a single line in the server configuration file.
```
Comment (pierce314159) [inline MODULAR.md:15]: small typo `buld process` -> `build process`

**Review by pierce314159 — APPROVED** [review]

File: MODULAR.md (hunk 4 above)
Comment (bmcdonald3) [reply MODULAR.md]: Good catch; updated.

**Comment by bmcdonald3** [comment]
Alright, updated this PR to use an environment variable named `ARKOUDA_SERVER_USER_MODULES` rather than requiring a full path in the `ServerModules.cfg` file.

File: src/serverModuleGen.py [hunk 5]
```diff
@@ -5,9 +5,12 @@ def getModules(filename):
...
         ret = []
         for module in modules:
-            module = (module.rstrip()).lstrip()
+            module = module.strip()
             if len(module) > 0 and module[0] != '#':
-                ret.append(module)
+                if module[0] == '/':
+                    ret.append(module[module.rfind('/')+1:])
+                else:
+                    ret.append(module)
```
Comment (glitch) [inline src/serverModuleGen.py:13]: This works so no change requested, but if you wanted to tighten it up a bit
```python
    for module in modules:
        module = module.split("#")[0].split("/")[-1].strip()
        if module:
            ret.append(module)
```
This also handles trailing comments i.e. `MyMsg  # This is my fancy module` -> `MyMsg`

If you want to go full blown crazy the whole func in 3 lines:
```python
def getModules(filename):
    with open(filename) as configfile:
        return [m.split("/")[-1].strip() for m in configfile.readlines() if m.split("#")[0].strip()]
```
but that is a bit unreadable.

**Review by glitch — APPROVED** [review]

File: src/serverModuleGen.py (hunk 5 above)
Comment (bmcdonald3) [reply src/serverModuleGen.py]: Ooh, I like the fancy Python syntax, switched to the first option, thanks!

**Pull request merged**